keps/2724-topology-aware-schedling/README.md

```diff
@@ -0,0 +1,569 @@
+global uniqueness of labels on each level, i.e. there could be two nodes with
+the same "rack" label, but in different "blocks".
+
+For example, this is a representation of the dataset hierarchy;
+
+|  node  |  cloud.provider.com/topology-block | cloud.provider.com/topology-rack |
+|:------:|:----------------------------------:|:--------------------------------:|
+| node-1 |               block-1              |              rack-1              |
+| node-2 |               block-1              |              rack-2              |
+| node-3 |               block-2              |              rack-1              |
+| node-4 |               block-2              |              rack-3              |
+
+Note that, there is a pair of nodes, node-1 and node-3, with the same value of
+the "cloud.provider.com/topology-rack" label, but in different blocks.
+
+### Admin-facing API
+
+```golang
+// ResourceFlavorSpec defines the desired state of the ResourceFlavor
+type ResourceFlavorSpec struct {
+    ...
+
+  // TopologyName indicates the name of the topology for the ResourceFlavor.
+  // When specified, it enables scraping of the topology information from the
+  // nodes matching to the Resource Flavor node labels.
+  TopologyName *string
+}
+
+// TopologySpec defines the desired state of Topology
+type TopologySpec struct {
+	// Levels defines the levels of topology.
+	Levels []TopologyLevel
+}
+
+// TopologyLevel defines the desired state of TopologyLevel
+type TopologyLevel struct {
+	// NodeLabel indicates the name of the node label for a specific topology
+	// level. Examples:
+  // - cloud.provider.com/topology-block
+  // - cloud.provider.com/topology-rack
+	NodeLabel string
+}
+```
+
+Example TAS Resource Flavor & Topology configuration:
+
+```yaml
+kind: ResourceFlavor
+metadata:
+  name: "tas-flavor"
+spec:
+  nodeLabels:
+    "cloud.provider.com/node-group: tas"
+  topologyName: default
+—k
+kind: Topology
+metadata:
+  name: "default"
+spec:
+  levels:
+  - nodeLabel: cloud.provider.com/topology-block
+  - nodeLabel: cloud.provider.com/topology-rack
+```
+
+### User-facing API
+
+The user will need to point the workload to the ClusterQueue with the TAS
+ResourceFlavor, and add one of the annotations:
+
+```golang
+const (
+
+  // This annotation indicates that a workload requires Topology Aware Scheduling,
+  // and running all pods on nodes closely connected within the same level of
+  // hierarchy is a strong requirement for scheduling the workload.
+  RequireTopologyAnnotation = "kueue.x-k8s.io/require-topology"
+
+  // This annotation indicates that a workload requires Topology Aware Scheduling,
+  // but running all pods without the same topology level is a preference rather
+  // than requirement. There is a distinguished value "auto" which means that
+  // the lowest hierarchy level should be used.
+  PreferTopologyAnnotation = "kueue.x-k8s.io/prefer-topology"
+)
+```
+
+### Validation
+
+We introduce the following validations:
```

Comment:
when is this validated?
I suppose at admission time, because it wouldn't be feasible at creation time.

Reply:
Added explanations. I think it cannot be creation time, but we can do prior to admission.